### PR TITLE
DOC: Timestamp EX01 errors

### PR DESCRIPTION
**Pull request by lucasrodes**
Example section added for several methods in class [pd.Timestamp](https://github.com/pandas-dev/pandas/blob/1f89b82cc65f512e78b0038975425cc3d0772a05/pandas/_libs/tslibs/timestamps.pyx#L768)

- [X] xref #37875
- [ ] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

---
Examples added for:

- pandas.Timestamp.asm8
- pandas.Timestamp.day_of_week
- pandas.Timestamp.day_of_year
- pandas.Timestamp.days_in_month
- pandas.Timestamp.is_leap_year
- pandas.Timestamp.is_month_end
- pandas.Timestamp.is_month_start
- pandas.Timestamp.is_quarter_end
- pandas.Timestamp.is_quarter_start
- pandas.Timestamp.is_year_end
- pandas.Timestamp.is_year_start
- pandas.Timestamp.quarter
- pandas.Timestamp.tz
- pandas.Timestamp.week
- pandas.Timestamp.ceil
- pandas.Timestamp.combine
- pandas.Timestamp.day_name
- pandas.Timestamp.floor
- pandas.Timestamp.fromordinal
- pandas.Timestamp.fromtimestamp
- pandas.Timestamp.month_name
- pandas.Timestamp.normalize
- pandas.Timestamp.now
- pandas.Timestamp.replace
- pandas.Timestamp.round
- pandas.Timestamp.strftime
- pandas.Timestamp.timestamp
- pandas.Timestamp.to_numpy
- pandas.Timestamp.to_julian_date
- pandas.Timestamp.to_period
- pandas.Timestamp.to_pydatetime
- pandas.Timestamp.today
- pandas.Timestamp.tz_convert
- pandas.Timestamp.tz_localize
- pandas.Timestamp.utcfromtimestamp
- pandas.Timestamp.utcnow
- pandas.Timestamp.weekday

### CODE REVIEW

**Comment by lucasrodes** [comment]
From the checks, I believe that there are some docstrings that were not supposed to be changed? 

I am starting to contribute and trying to follow up on this #17327

File: pandas/_libs/tslibs/timestamps.pyx [hunk 1]
```diff
@@ -737,12 +871,30 @@ cdef class _Timestamp(ABCTimestamp):
...
 
     def to_period(self, freq=None):
         """
         Return an period of which this timestamp is an observation.
+
+        Examples
+        --------
+        >>> ts = pd.Timestamp('2020-03-14T15:32:52.192548651')
+        >>> ts.to_period(freq='Y) # Year end frequency
+        numpy.datetime64('2020-03-14T15:32:52.192548651')
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx:891]: can you put a blank line between cases

File: pandas/_libs/tslibs/timestamps.pyx [hunk 2]
```diff
@@ -1162,6 +1358,37 @@ timedelta}, default 'raise'
...
+        Create a timestamp object:
+
+        >>> ts = pd.Timestamp('2020-03-14T15:32:52.192548651')
+
+        A timestamp can be rounded using multiple frequency units:
+
+        >>> ts.round(freq='H') # hour
+        Timestamp('2020-03-14 16:00:00')
+        >>> ts.round(freq='T') # minute
+        Timestamp('2020-03-14 15:33:00')
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx:1373]: can you put a blank line between cases

File: pandas/_libs/tslibs/timestamps.pyx [hunk 3]
```diff
@@ -1203,6 +1430,37 @@ timedelta}, default 'raise'
...
+        --------
+        Create a timestamp object:
+
+        >>> ts = pd.Timestamp('2020-03-14T15:32:52.192548651')
+
+        A timestamp can be floored using multiple frequency units:
+
+        >>> ts.floor(freq='H') # hour
+        Timestamp('2020-03-14 15:00:00')
+        >>> ts.floor(freq='T') # minute
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx:1444]: same

File: pandas/_libs/tslibs/timestamps.pyx [hunk 4]
```diff
@@ -1203,6 +1430,37 @@ timedelta}, default 'raise'
...
+        Timestamp('2020-03-14 15:32:00')
+        >>> ts.floor(freq='S') # seconds
+        Timestamp('2020-03-14 15:32:52')
+        >>> ts.floor(freq='L') # miliseconds
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx:1448]: prob can show just a few rather than all

File: pandas/_libs/tslibs/timestamps.pyx [hunk 5]
```diff
@@ -1242,13 +1500,50 @@ timedelta}, default 'raise'
...
+
+        >>> ts = pd.Timestamp('2020-03-14T15:32:52.192548651')
+
+        A timestamp can be ceiled using multiple frequency units:
+
+        >>> ts.ceil(freq='H') # hour
+        Timestamp('2020-03-14 16:00:00')
+        >>> ts.ceil(freq='T') # minute
+        Timestamp('2020-03-14 15:33:00')
+        >>> ts.ceil(freq='S') # seconds
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx:1516]: same comment

**Review by jreback — CHANGES_REQUESTED** [review]
looks good a couple of comments. pls merge master and ping on green.

**Comment by jreback** [comment]
can you merge master and ping on green (to make sure no failures are real).

**Comment by lucasrodes** [comment]
Added few changes to [pandas/_libs/tslibs/nattype.pyx](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/nattype.pyx) so the docstrings match. Should pass the tests now, I'll ping you again. specially this one below:

https://github.com/pandas-dev/pandas/blob/456fcb9829f4e8fba200658efff6ba26152009cd/pandas/tests/scalar/test_nat.py#L326

**Comment by lucasrodes** [comment]
@jreback passing all tests

File: pandas/_libs/tslibs/timestamps.pyx (hunk 2 above)
Comment (lucasrodes) [reply pandas/_libs/tslibs/timestamps.pyx:1377]: Done

File: pandas/_libs/tslibs/timestamps.pyx (hunk 1 above)
Comment (lucasrodes) [reply pandas/_libs/tslibs/timestamps.pyx:891]: Done

File: pandas/_libs/tslibs/timestamps.pyx (hunk 3 above)
Comment (lucasrodes) [reply pandas/_libs/tslibs/timestamps.pyx:1447]: Done

File: pandas/_libs/tslibs/timestamps.pyx (hunk 5 above)
Comment (lucasrodes) [reply pandas/_libs/tslibs/timestamps.pyx:1519]: Done

File: pandas/_libs/tslibs/nattype.pyx [hunk 6]
```diff
@@ -252,6 +252,12 @@ cdef class _NaT(datetime):
         See Also
         --------
         DatetimeIndex.to_numpy : Similar method for DatetimeIndex.
+
+        Examples
```
Comment (jreback) [inline pandas/_libs/tslibs/nattype.pyx:256]: these need to be NaT examples (same for anything in nattype)

**Review by jreback — CHANGES_REQUESTED** [review]
ok happy for you to revert the nattype updates and we can address in another PR. ping on green.

**Comment by lucasrodes** [comment]
Sure @jreback, can do that. However, this pull request won't go through the tests as I mentioned before (won't be green). Should we merge it regardless?

**Comment by lucasrodes** [comment]
@jreback changes reverted

**Comment by jreback** [comment]
> Sure @jreback, can do that. However, this pull request won't go through the tests as I mentioned before (won't be green). Should we merge it regardless?

oh how is master passing now then?

yeah this needs to be green. ok i guess re-revert. the only thing with the NaT tests is that they should be showing NaT int he examples

**Comment by lucasrodes** [comment]
Hi @jreback ,
The test I attach below is checking if the docstrings between methods in  [pandas/_libs/tslibs/nattype.pyx](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/nattype.pyx) and  [pandas/_libs/tslibs/timestamps.pyx](pandas/_libs/tslibs/timestamps.pyx) match. That's why I also used pd.Timestamp instead of pd.NaT.

https://github.com/pandas-dev/pandas/blob/456fcb9829f4e8fba200658efff6ba26152009cd/pandas/tests/scalar/test_nat.py#L326


 Should we maybe change the test [pandas/pandas/tests/scalar/test_nat.py](pandas/pandas/tests/scalar/test_nat.py) in onther PR?

**Comment by jreback** [comment]
> Hi @jreback ,
> The test I attach below is checking if the docstrings between methods in [pandas/_libs/tslibs/nattype.pyx](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/nattype.pyx) and [pandas/_libs/tslibs/timestamps.pyx](pandas/_libs/tslibs/timestamps.pyx) match. That's why I also used pd.Timestamp instead of pd.NaT.
> 
> https://github.com/pandas-dev/pandas/blob/456fcb9829f4e8fba200658efff6ba26152009cd/pandas/tests/scalar/test_nat.py#L326
> 
> Should we maybe change the test [pandas/pandas/tests/scalar/test_nat.py](pandas/pandas/tests/scalar/test_nat.py) in onther PR?

ok instead let's use the same doc-string, BUT add a NaT example to it, this should make tests pass and make the examples useful.

ping on green.

**Comment by lucasrodes** [comment]
test

**Comment by lucasrodes** [comment]
Error in the following test from [pandas/tests/io/test_gcs.py](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/io/test_gcs.py) in Windows. Not sure what is causing it.


https://github.com/pandas-dev/pandas/blob/5abc06fbb9b9f7cf63059695bf7fa87d9b57df5c/pandas/tests/io/test_gcs.py#L83-L119

**Comment by lucasrodes** [comment]
On another note:

Hi @jreback, thanks for your feedback.

Let me summarize last commits below.

### 1) `NaT` examples added
I have added examples with NaT in the docstrings of the following functions in [nattype.pyx](
https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/nattype.pyx) and [timestamps.pyx](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/timestamps.pyx):

- tz_convert
- tz_localize
- today
- to_pydatetime
- to_numpy
- round
- replace
- now
- month_name
- floor
- day_name
- ceil

Hence, **these methods now have the same docstring in both modules and should pass the tests** :heavy_check_mark:.

### 2) `NaT` examples not added
Next, I have not added `NaT` examples in the following examples as **these methods basically raise a `ValueError` if called from a `NaT` object**. However I have copied the docstrings from [timestamps.pyx](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/timestamps.pyx) (otherwise tests are not passed).

- utcnow
- utcfromtimestamp
- timestamp
- strftime
- fromtimestamp
- fromordinal
- combine
- astimezone

Note that these methods were created using method [`_make_error_func`](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/nattype.pyx#L67) instead of [`_make_nat_func`](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/nattype.pyx#L59) or [`_make_nan_func`](https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/tslibs/nattype.pyx#L51).

As **these methods now have the same docstring in both modules and should pass the tests** :heavy_check_mark:.

---

### 3) Way forward

From here I see three ways out:

A) Review of this PR, iterate and merge.
B) Remove the Example section from all methods listed in 2) in both modules (so docstrings still match), commit, review and merge PR.
C) Review the tests in [test_nat.pyx](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/scalar/test_nat.py). In particular the parametrization of `compare`. Probably another PR is required.

https://github.com/pandas-dev/pandas/blob/11a5a4b1b11ad68999b6d153cca99907db43affe/pandas/tests/scalar/test_nat.py#L319-L334

**Comment by jreback** [comment]
@lucasrodes i am ok prob merging this as is, if you can merge master and ping

**Comment by jreback** [comment]
@lucasrodes this was close. if you can merge master and see if can get all passing.

**Comment by lucasrodes** [comment]
Sorry for the late reply @jreback. Will do asap

**Comment by lucasrodes** [comment]
@jreback merged with master

File: pandas/_libs/tslibs/nattype.pyx [hunk 7]
```diff
@@ -787,6 +1047,24 @@ default 'raise'
...
+        Create a naive timestamp object:
+
+        >>> ts = pd.Timestamp('2020-03-14T15:32:52.192548651')
+        >>> ts
+        Timestamp('2020-03-14 15:32:52.192548651')
+
+        Add 'Europe/Stockholm' as timezone:
+
+        >>> ts.tz_localize(tz='Europe/Stockholm')
+        Timestamp('2020-03-14 15:32:52.192548651+0100', tz='Europe/Stockholm'
```
Comment (mroeschke) [inline pandas/_libs/tslibs/nattype.pyx:1062]: Missing an ending `)`

**Comment by mroeschke** [comment]
Sorry for the long delay @lucasrodes . Looks mostly good. One comment otherwise LGTM

**Comment by lucasrodes** [comment]
@mroeschke Merged with upstream & fixed typos (missing '(' in docstring examples)

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lucasrodes! Apologies for the long wait getting this one in.

**Pull request merged**